### PR TITLE
lapack/gonum: add Dgetc2

### PR DESCRIPTION
**Pull request by soypat**
In hopes of someday getting to add CARE solver for #1651, blas routine `dgetc2` is added.

I'd like some guidance on testing. I've looked at tests in the BLAS package and they're all quite different from what I've seen in the LAPACK package. What stands out to me particularily is the tables:

```go
for i, test := range []struct {
		ul   blas.Uplo
		tA   blas.Transpose
		d    blas.Diag
		n, k int
		a    [][]float64
		x    []float64
		incX int
		ans  []float64
	}{
		{
			ul: blas.Upper,
			tA: blas.NoTrans,
			d:  blas.NonUnit,
			n:  5,
			k:  1,
			a: [][]float64{
				{1, 3, 0, 0, 0},
				{0, 6, 7, 0, 0},
				{0, 0, 2, 1, 0},
				{0, 0, 0, 12, 3},
				{0, 0, 0, 0, -1},
			},
			x:    []float64{1, 2, 3, 4, 5},
			incX: 1,
			ans:  []float64{2.479166666666667, -0.493055555555556, 0.708333333333333, 1.583333333333333, -5.000000000000000},
		},
// ...
```
Is this the way to go?

### CODE REVIEW

**Comment by vladimir-ch** [comment]
This should go into `lapack/{gonum,testlapack}`, it's not a BLAS routine.

The test should follow the `lapack/testlapack` style, for example `dpbtf2.go` is kind of related and relatively recent. `dtrtrs.go` is not so related but also quite recent and shows the style on which we've recently settled in lapack.

**Comment by soypat** [comment]
Got confused since the page says "BLAS Level 2 *Algorithm*". will fix.

Thanks for the suggestions!

File: blas/gonum/level2float64.go
Comment (vladimir-ch) [inline blas/gonum/level2float64.go]: Gonum is row-major (rows are continuous in storage) while Fortran is column-major (columns are continuous in storage). So Lapack must use increment of `lda` to get to the next element in a row but for us the increment is 1. Below, where columns are swapped, the situation is reversed, Lapack increments by 1 to get to the next element in a column but we need to increment by `lda`. Therefore here:
```
impl.Dswap(n, a[ipv*lda:], 1, a[i*lda:], 1)
```

File: blas/gonum/level2float64.go
Comment (vladimir-ch) [inline blas/gonum/level2float64.go]: `i*lda+i` (the column index is `i`, not 1)

File: blas/gonum/level2float64.go
Comment (vladimir-ch) [inline blas/gonum/level2float64.go]: Also here.

File: blas/gonum/level2float64.go
Comment (vladimir-ch) [inline blas/gonum/level2float64.go]:
```suggestion
		impl.Dger(n-i-1, n-i-1, -1.0, a[(i+1)*lda+i:], lda, a[i*lda+i+1:], 1, a[(i+1)*lda+i+1:], lda)
```

File: blas/gonum/level2float64.go
Comment (vladimir-ch) [inline blas/gonum/level2float64.go]: Here `n` is used as an index, so we must subtract 1:
```suggestion
	if math.Abs(a[(n-1)*lda+n-1]) < smin {
```

File: blas/gonum/level2float64.go
Comment (vladimir-ch) [inline blas/gonum/level2float64.go]:
```suggestion
		k = n-1
```

File: blas/gonum/level2float64.go
Comment (vladimir-ch) [inline blas/gonum/level2float64.go]:
```suggestion
		a[(n-1)*lda+n-1] = smin
```

File: blas/gonum/level2float64.go
Comment (vladimir-ch) [inline blas/gonum/level2float64.go]:
```suggestion
	ipiv[n-1], jpiv[n-1] = n-1, n-1
```

File: blas/gonum/level2float64.go
Comment (vladimir-ch) [inline blas/gonum/level2float64.go]:
```suggestion
	return k
```

File: blas/gonum/level2float64.go
Comment (vladimir-ch) [inline blas/gonum/level2float64.go]: Delete empty line.

File: blas/gonum/level2float64.go
Comment (vladimir-ch) [inline blas/gonum/level2float64.go]: Delete empty line.

File: blas/gonum/level2float64.go
Comment (vladimir-ch) [inline blas/gonum/level2float64.go]: Delete empty line.

File: blas/gonum/level2float64.go
Comment (vladimir-ch) [inline blas/gonum/level2float64.go]:
```suggestion
			k = 0
		}
		return k
```

File: blas/gonum/level2float64.go
Comment (vladimir-ch) [inline blas/gonum/level2float64.go]: Move this before quick return.

File: blas/gonum/level2float64.go
Comment (vladimir-ch) [inline blas/gonum/level2float64.go]:
```suggestion
		return k
```

**Review by vladimir-ch — CHANGES_REQUESTED** [review]
Let's see what Github does with these review comments done when the code was still in the blas directory.

**Comment by vladimir-ch** [comment]
> Got confused since the page says "BLAS Level 2 Algorithm". will fix.

No worries. Some Lapack routines come in 2 variants, a simpler one that uses BLAS Level 2 and operates on columns/rows, and a more complex one that uses BLAS Level 3, operates on blocks and should be more performant. The comment here means that this routine _uses_ BLAS Level 2.

**Comment by vladimir-ch** [comment]
> Let's see what Github does with these review comments done when the code was still in the blas directory.

Ok, it shows them as outdated but they are of course all still valid.

File: lapack/gonum/dgetc2.go
Comment (vladimir-ch) [inline lapack/gonum/dgetc2.go:52]:
```suggestion
		ipiv[0], jpiv[0] = 0, 0
```

File: blas/gonum/level2float64.go
Comment (soypat) [reply blas/gonum/level2float64.go]: darn, missed that one

File: blas/gonum/level2float64.go
Comment (soypat) [reply blas/gonum/level2float64.go]: Think you may have missed the last `lda` which should actually be a `1`, right?

Comment (soypat) [reply blas/gonum/level2float64.go]: Or no, that one **is** equal to `lda` since it is `lda` parameter that is passed into `Dger`

File: blas/gonum/level2float64.go
Comment (soypat) [reply blas/gonum/level2float64.go]: Woah, slipped right past me. Thanks Vladimir.

**Comment by soypat** [comment]
I'm at a loss as to how to construct the permutation matrices from pivot indices. Cholesky factorization function does not have a similar implementation.

**Comment by kortschak** [comment]
Why do you need to construct a permutation matrix?

File: blas/gonum/level2float64.go
Comment (vladimir-ch) [reply blas/gonum/level2float64.go]: `x` is i-th column of A, so Lapack has `inc=1`, we have `inc=lda`, `y` is (i+1)-th row of A, Lapack has `inc=lda`, we have `inc=1`. The last argument is the destination matrix, so `lda` is kept.

File: lapack/gonum/dgetc2.go
Comment (vladimir-ch) [inline lapack/gonum/dgetc2.go:79]: Move this out of the for loop, before `var smin float64`.

File: lapack/gonum/dgetc2.go
Comment (vladimir-ch) [inline lapack/gonum/dgetc2.go:49]: nit: I think this comment is not so useful and I'd remove it.

File: lapack/gonum/dgetc2.go
Comment (vladimir-ch) [inline lapack/gonum/dgetc2.go:108]: Either remove the comment or change `n` -> `n-1`.

**Review by vladimir-ch — CHANGES_REQUESTED** [review]
`Dgetc2` looks almost ok now, I haven't looked at the test because you're still working on it.

**Comment by soypat** [comment]
@kortschak I'm way out of my element here, but I'd though a good way to test `Dgetc2` was to reconstruct the matrix `A` from the results of the LU factorization `A = P * L * U * Q` where P and Q are permutation matrices, L is lower triangular with unit diagonal elements and U is upper triangular.

**Comment by vladimir-ch** [comment]
> I'd thought that a good way to test Dgetc2 was to reconstruct the matrix A from the results of the LU factorization A = P * L * U * Q where P and Q are permutation matrices, L is lower triangular with unit diagonal elements and U is upper triangular.

Yes, reconstructing A is the way to test Dgetc2 but you don't need to form P and Q explicitly and multiply, you can use ipiv and jpiv to swap rows/columns to achieve the same effect.

> What my test is doing thus far is checking if the permutation indices yield unique combinations, which I'm not even sure is an expected result. It is failing for now.

Testing the content of ipiv and jpiv is easy and will not hurt. I would fill them with negative numbers before calling Dgetc2, then make their copies, sort the copies and then check that ipivCopy[i] == i for all i. Then I would use ipiv, jpiv to reconstruct A.

**Comment by soypat** [comment]
> Testing the content of ipiv and jpiv is easy and will not hurt. I would fill them with negative numbers before calling Dgetc2, then make their copies, sort the copies and then check that ipivCopy[i] == i for all i. Then I would use ipiv, jpiv to reconstruct A.

I got around to implementing this and letting it fail. I've read the code and compared it with reference but nothing seems out of place. Any ideas on what I should be looking out for?

**Comment by vladimir-ch** [comment]
That's my fault, I didn't pay enough attention to how the permutations are represented in ipiv and jpiv. Now I see there is not much that can be tested except that all elements of ipiv, jpiv are set, in your case that no element is -1. After that you should focus on checking that A can be reconstructed from the factorization.

**Comment by soypat** [comment]
>  you can use ipiv and jpiv to swap rows/columns to achieve the same effect.
There's definitely more to reconstructing A than swapping columns and rows. I did just that in my test but noticed the lines 
```go
for j := i + 1; j < n; j++ {
			a[j*lda+i] /= a[i*lda+i]
		}
		bi.Dger(n-i-1, n-i-1, -1.0, a[(i+1)*lda+i:], lda, a[i*lda+i+1:], 1, a[(i+1)*lda+i+1:], lda)
```
modify `a` in every case. There's gotta be a better way of testing this than programming `inverseDgetc2`.

**Comment by vladimir-ch** [comment]
> There's definitely more to reconstructing A than swapping columns and rows.

Yes, Dgetc2 stores the L and U triangular factors in-place into a (and P and Q into ipiv and jpiv).

**Comment by soypat** [comment]
@kortschak This branch is ready to merge. `Dgetc2` passes all tests.

File: lapack/gonum/dgetc2.go
Comment (kortschak) [inline lapack/gonum/dgetc2.go:47]:
```
const (
    eps    = dlamchP
    smlnum = dlamchS / eps
)
```

File: lapack/gonum/dgetc2.go
Comment (kortschak) [inline lapack/gonum/dgetc2.go:22]: Dgetc2 is auxiliary, so note that it's only exported for testing purposes.

```
//
// Dgetc2 is an internal routine. It is exported for testing purposes.
```

File: lapack/gonum/dgetc2.go
Comment (kortschak) [inline lapack/gonum/dgetc2.go:14]: s/n-b-n/n×n/

File: lapack/gonum/dgetc2.go
Comment (kortschak) [inline lapack/gonum/dgetc2.go:21]: This doesn't look like our normal style of docs. I'll let @vladimir-ch comment on this.

File: lapack/gonum/dgetc2.go
Comment (kortschak) [inline lapack/gonum/dgetc2.go:59]: lc/SMIN/g

File: lapack/gonum/dgetc2.go
Comment (kortschak) [inline lapack/gonum/dgetc2.go:69]: I'm not sure that this is clearer on a single line. ¯\\\_ (ツ)\_/¯

File: lapack/gonum/dgetc2.go
Comment (kortschak) [inline lapack/gonum/dgetc2.go:98]: s/-1.0/-1/

File: lapack/gonum/dgetc2.go
Comment (kortschak) [inline lapack/gonum/dgetc2.go:107]: Again, I think this probably should be two lines.

Comment (kortschak) [inline lapack/gonum/dgetc2.go:107]: All this get simpler if there is a `n--` at the beginning. Too clever, @vladimir-ch?

File: lapack/testlapack/dgetc2.go
Comment (kortschak) [inline lapack/testlapack/dgetc2.go:1]: 2021

File: lapack/testlapack/dgetc2.go
Comment (kortschak) [inline lapack/testlapack/dgetc2.go:13]: Blank line between us and them.

File: lapack/testlapack/dgetc2.go
Comment (kortschak) [inline lapack/testlapack/dgetc2.go:29]: Capitalisation and final dot.

File: lapack/testlapack/dgetc2.go
Comment (kortschak) [inline lapack/testlapack/dgetc2.go:81]: Final dot.

File: lapack/gonum/dgetc2.go
Comment (soypat) [reply lapack/gonum/dgetc2.go]: I've looked at `Dgelq2` to make the documentation better for last commit.

File: lapack/testlapack/dgetc2.go
Comment (vladimir-ch) [inline lapack/testlapack/dgetc2.go:84]:
```suggestion
	lu := cloneGeneral(a)
```

File: lapack/testlapack/dgetc2.go
Comment (vladimir-ch) [inline lapack/testlapack/dgetc2.go:87]: I'm not sure this deserves a `Fatalf`. It's not an indication of a bug, it's about the input data. We're not hitting it but please change to `Logf`.

File: lapack/testlapack/dgetc2.go
Comment (vladimir-ch) [inline lapack/testlapack/dgetc2.go:118]: This will break once `lda != n`. Either use `ldl := n` etc or use `zeros` to allocate `L`, `U`, `work` and use their `Stride` field here.

File: lapack/testlapack/dgetc2.go
Comment (vladimir-ch) [inline lapack/testlapack/dgetc2.go:132]: This again assumes that `lu` (`a`) and `work` slices have the same length which will not be true when `lda != n`.

**Review by vladimir-ch — CHANGES_REQUESTED** [review]
Also please remove `lapack/.gitignore`.

File: lapack/gonum/dgetc2.go
Comment (vladimir-ch) [reply lapack/gonum/dgetc2.go]: It's ok. I'll polish it after it's merged.

File: lapack/testlapack/dgetc2.go
Comment (soypat) [reply lapack/testlapack/dgetc2.go:36]: Those are test cases I created by running LAPACK's `Dgetc2` and copying the output as `expect`. `"identity"` test case is supposed to be a property based test: the identity matrix is already LU.

File: lapack/testlapack/dgetc2.go
Comment (soypat) [reply lapack/testlapack/dgetc2.go:39]: I chose an `A` matrix that had a rational LU decomposition. I'm not sure what you mean by the second sentence.

File: lapack/testlapack/dgetc2.go
Comment (soypat) [reply lapack/testlapack/dgetc2.go:30]: Only the Identity case which I thought of a special property based test. Though when you say it like that I guess there's not much added by the tests. We can remove them

File: lapack/testlapack/dgetc2.go
Comment (vladimir-ch) [reply lapack/testlapack/dgetc2.go:39]: I mean that both 6 are equally good candidates for the first pivot and there is nothing in the docs that would say which one will be chosen. In other words, the test expectation depends on an implementation detail.

File: lapack/testlapack/dgetc2.go
Comment (vladimir-ch) [reply lapack/testlapack/dgetc2.go:36]: Do you mean that you created the output by running your implementation of Dgetc2 or the Fortran reference? Clearly, you can't generate the expected output by the very implementation that you're testing.

My question about ipiv, jpiv again points to the ambiguity how the pivots are chosen by the algorithm when there are more candidates.

File: lapack/testlapack/dgetc2.go
Comment (vladimir-ch) [reply lapack/testlapack/dgetc2.go:30]: Yes, I think it's ok to remove them.

Comment (soypat) [reply lapack/testlapack/dgetc2.go:30]: You've made your point clear as day. Thanks for clearing the reason up!

**Comment by soypat** [comment]
Changes to tests done as per your suggestion. Removed those cases generated by the reference since they may fail depending on implementation details.

**Review by vladimir-ch — APPROVED** [review]
LGTM, thank you for your contribution and following our guidance. I will squash the commits before merging and then submit a PR cleaning up a few nits.

I think this is your first contribution; @kortschak , what is the process?

**Comment by vladimir-ch** [comment]
I found it `CONTRIBUTING.md`:

> you will need to be added to the CONTRIBUTORS and AUTHORS file; after your contribution has been accepted you will be asked to open a pull request adding yourself to them.

@soypat , can you submit a new PR where you add yourself to those two files?

**Comment by kortschak** [comment]
Make the commit message "A+C: add Patricio Whittingslow".

**Pull request merged**
